Merge josephtheengineer/develop

josephtheengineer opened the pull request:
commit 22813099357431257c4bc93904fc36f7fcfebfd6 (HEAD -> develop, origin/develop)
Author: josephtheengineer <joseph@theengineer.life>
Date:   Sun Jun 14 14:23:49 2020 +1000

    fix: logging now works again, renamed Characters
    
    Logging
    - Make the logger a preload var
    - Add logger.init() to _ready in Core
    - Add script_name in Log
    - Move Log message to after the signal gets setup
    
    CharacterDefaults.gd -> Characters.gd
    - Rename CharacterDefaults to Characters to keep
    in line with other names, eg, Attacks, Items
    
    CharacterCreation
    - Fixed a bug where after creating a character
      your stats would be set to an array rather then
      a Dictonary
    
    process_level_buffs
    - Now takes a character_class argument
    - process_level_buffs() is now ran when
      set_level() is run
    
    Fixed infinite loop bug
    - Put process_level_buffs count and temp_level changes
      inside the loop, woops

commit 148de882c225fb81b429af5eced56ef9aa184430 (HEAD -> develop, origin/develop)
Author: josephtheengineer <joseph@theengineer.life>
Date:   Sat Jun 13 16:41:52 2020 +1000

    refactor: move Core log to Log, rename BattleScene
    
    Log
    - Move Core functions responsible for logging to a
      seperate Log class
    - Change all references to Core.LEVEL to Log.LEVEL
    
    Battle
    - Rename BattleScenes folder to Battle because the
      parent is also call scenes
    
    Starting Equipment
    - Make items a dictonary and allow a quantity field
    - This allows us to easily give the the player 15
      arrows with 'quantity = 15'
    - Reflect this change in CharacterManager
    
    CharacterManager
    - Implement set_level and process_level
    
    InventoryManager
    - Fix an issue where items would no get their quantity
      set if the item doesn't already exist
    
      add 'item_data.quantity = quantity'
    - Fix an issue where items would check different just
      becuase they have a different quantity
    
      Solution: remove the quantity value before comparing
                item_check_data.erase("quantity")
    
    ItemManager
    - Move function to InventoryManager and delete

commit 6d2b1fca3c5dc2cb3a2e3360faf9d76b0c217907 (HEAD -> develop, origin/develop)
Author: josephtheengineer <joseph@theengineer.life>
Date:   Sat Jun 13 12:09:50 2020 +1000

    feature: add wooden_ui assets, fog and sound fx
    
    - Add Assets/Images/Wooden_UI
    - Add various Assets/Sounds/Effects
    - Add various Assets/Sounds/Music
    
    AudioManager
    - Create AudioManager for easily creating audio
      players
    
    CharacterManager.gd
    - remove unneeded whitespace
    
    AudioDebugScene
    - Create AudioDebugScene for testing the audio buses
    
    CharacterCreation.gd
    - add support for flashing text button
    
    CharacterCreation.tscn
    - Add fog background
    - Make button text flash when enabled
    - Change GAME_TITLE -> RPG GODOT
    
    CharacterSelection.gd
    - Add support for a profile border
    - Fix crash when selecting a character
      - change saveFile -> file
    
    CharacterSelection.tscn
    - Make selection buttons different colors (Wooden_UI/frame)
    - Add banner to the top of the scene where the title is
    
    CharacterSelectionButton.tscn
    - Add profile border
    - Change button background to Wooden_UI/bg_01_02_landscape.png
    
    FlashingText
    - Create text script and scene that flashes when
      state is set to States.FLASHING

    Loading
    - Make loading show the script_name and level like in Core
    - Add 1 sec wait
    - Make loading play Music via the AudioManager
    
    Overlay
    - Create an Overlay / Background element that contains
      some cool animations
    - Overlay runs lightning storm on splash screen
    - The lightning storm uses a 2d light and some sound effects
    - Overlay animates the fog
    - Overlay renders some 2d particles Over top the screen for
      a cool depth effect
    
    Splash
    - Add script name to splash
    - Make splash call start-storm on Overlay
    - Splash tells the flashing text to show after the logo
      text has faded in
    
    Shaders/noise.shader
    - Add generic noise shader (used for fog)
    
    AudioBus
    - Create default_bus_layout.tres to seperate and control the
      volume of different sounds
    - Some buses also have different sound effects / filters
    
    layer_names
    - Add layers for 2d physics

commit e001774df0282e4ef4245d6efa51b8e6e8d3fa77 (HEAD -> develop, origin/develop)
Author: josephtheengineer <joseph@theengineer.life>
Date:   Fri Jun 12 19:36:56 2020 +1000

    refactor: remove unused global class Battle

commit 5490a8f4f154557c5ef7d75989984fdc70df0d3d (HEAD -> develop, origin/develop)
Author: josephtheengineer <joseph@theengineer.life>
Date:   Fri Jun 12 19:27:57 2020 +1000

    fix: equiped items are now stored by their index
    
    - Change var characters to const characters in CharacterDefaults
    - Refactor CharacterManager
    - Make InventoryManager work on a specific inventory and have
    CharacterManager wrap that
    
Fixes #78
Works on #77 #76

commit fe93db748a25a1833237f1f35ab6b86dbdca719b
Author: josephtheengineer <joseph@theengineer.life>
Date:   Fri Jun 12 18:26:29 2020 +1000

    feature: add a function to learn attacks
    
    - Replace character.attacks[attack_type].append(attack) to
      learn_attack(character_data, attack_type, attack_name)
    - Also rename character -> character_data in load_class
    - Also remove unneeded argument 'item' from calcuate_item_buffs()

commit d15061d38c7f4b452611f5fe3e5457b7dc30d1ae (HEAD -> develop, origin/develop, github/develop)
Author: josephtheengineer <joseph@theengineer.life>
Date:   Fri Jun 12 17:36:31 2020 +1000

    refactor: hard code in class attributes
    
    - Add const starting_attacks to CharacterDefaults.gd
    - Item buffs are now calcuated when needed and not stored (no reduct
    complexity)
    - Items and Inventories are now stored as pure data
    - Functions moved to InventoryManager and ItemManager
    - Refactored InventoryManager
    - Updated item buffs to correct values
    - Implement function load_class(character, character_class)
    - Changed 'weapon' type for items to a more general 'equipment' type
    - Replaced CharacterCreation's match chosen_equip with
    CharacterManager.load_class(player, chosen_equip)

commit ff7fbe4217bcddaf5a5a6fa0df181ee66358cd45 (HEAD -> develop, origin/develop, github/develop)
Author: josephtheengineer <joseph@theengineer.life>
Date:   Fri Jun 12 12:18:35 2020 +1000

    fix: item equip now works again
    
    - Reverted the item adding changes to CharacterCreation.gd
    - Changed Item _init buffs to a Dictionary

commit 9c642834ad028eedfcecf2ba068896b4c96a3a4a (HEAD -> develop, origin/develop, github/develop)
Author: josephtheengineer <joseph@theengineer.life>
Date:   Fri Jun 12 11:01:09 2020 +1000

    fix: profile images do not load
    
    - Add "res://Assets/Images/Profiles/" to picture path
    
Fixes #74

commit 1b52fc6b4f9fb3673eeec03cf0907d04596dd7e3
Author: josephtheengineer <joseph@theengineer.life>
Date:   Fri Jun 12 10:45:46 2020 +1000

    revert: fix inventory items
    
    - Add Item.gd
    - Add Inventory.gd

commit 60c65606573ce7618a7768f3ad48ab63f759002f (HEAD -> develop, origin/develop)
Author: josephtheengineer <joseph@theengineer.life>
Date:   Thu Jun 11 10:40:17 2020 +1000

    fix: get items working with new system
    
    - Update print to Core.emit_signal
    - Change (ex) Attacks.melee_attacks to Attacks.melee
    - Fix func update_attacks(CharacterIndex):
    - Fix print error on ok save_character(data) make_dir
    - Remove if !data.has("player"): from load_character since
    file structure has changed
    
Fixes #73

commit cdba88b0b3e4841065408b8e10f5abb0680bd2fe (HEAD -> develop, origin/develop)
Author: josephtheengineer <joseph@theengineer.life>
Date:   Thu Jun 11 07:18:56 2020 +1000

    refactor: change the way characters are created
    
    - Create Classes/Attacks.gd to store attacks
    - Create Classes/CharacterDefaults.gd to store default character stats
    - Move Classes/Characters/Character.gd functions to
    Classes/CharacterManager.gd and make them static
    - Created Classes/DictonaryFunc.gd for useful dictonary functions
    - Created Classes/Items.gd to store item stats
    - Updated Battle.gd, CharacterCreation and CharacterSelection to work with changes
    - SaveGame.gd: Removed character_class_to_data(Character) and
    data_to_character_class(player) since this is no longer needed

commit df2e181b1a648622213dea54efe6628cf4375ad4 (HEAD -> develop, origin/develop, github/develop)
Author: josephtheengineer <joseph@theengineer.life>
Date:   Fri Jun 5 23:02:06 2020 +1000

    refactor: lots of cleanup and add splash screen
    
    - Add flashing text scene in Scenes/FlashingText/
    - Add splash scene in Scenes/Splash/
    - Moved global classes to Classes/
    - Implement global scene loading event
    - Add music player
    
Fixes #53 
Fixes #71

commit 73046cdcd1769691288829ca23aabd34da81f177
Author: josephtheengineer <joseph@theengineer.life>
Date:   Fri Jun 5 22:57:37 2020 +1000

    feature: add some music and sound effects
    
    - All music from FreeSound.org and is (CC BY 3.0)
    - Added sound effects for looting and stabbing
    - Added background music
    - Added short suspense music
    
Fixes #70

commit aa3688621298d4a69858f44d2ab680472eacf820 (HEAD -> develop, origin/develop)
Author: josephtheengineer <joseph@theengineer.life>
Date:   Thu Jun 4 11:11:43 2020 +1000

    fix: game now checks if a save file is invalid
    
    - Added check in load_character to return false
    - load_all no longer adds saves with the value false
    
Fixes #61

commit 241b9ca0b33da7c605aa76de65311ca15e3ee391
Author: josephtheengineer <joseph@theengineer.life>
Date:   Thu Jun 4 11:48:15 2020 +1000

    feature: get attack log working
    
    - Added msg signal to friendly / enemy death
    - Added signal emit to show dealt damage
    - Added script_name to all scripts to easily identify where a message
    came from
    - Added function to pass msg signals to print() and save logs to
    latest.txt
    - Added function to add msg text to battle scene
    - Removed underscores in front of global signals (does not conform to
    style guide)
    
Fixes #63

commit 6c97f863abc28eb7a62b7263ff0e2de3ca4cf7c7
Author: josephtheengineer <joseph@theengineer.life>
Date:   Thu Jun 4 13:00:37 2020 +1000

    feature: add the ability to select a character
    
    - Move the select button so that it is infront of the other elements
    - Have the select button render behind it's parent
    - Change the select character button background color
    - Create a small select button interface script
    - Hook up play button signal
    - Add Vbox to SelectCharacter scene to organise nodes
    - Added selected_character var to select character
    
Fixes #67

Commit:
fix: game now checks if a save file is invalid

- Added check in load_character to return false
- load_all no longer adds saves with the value false

Fixes #61

Commit:
feature: get attack log working

- Added msg signal to friendly / enemy death
- Added signal emit to show dealt damage
- Added script_name to all scripts to easily identify where a message
came from
- Added function to pass msg signals to print() and save logs to
latest.txt
- Added function to add msg text to battle scene
- Removed underscores in front of global signals (does not conform to
style guide)

Fixes #63

Commit:
feature: add the ability to select a character

- Move the select button so that it is infront of the other elements
- Have the select button render behind it's parent
- Change the select character button background color
- Create a small select button interface script
- Hook up play button signal
- Add Vbox to SelectCharacter scene to organise nodes
- Added selected_character var to select character

Fixes #67

Commit:
feature: add some music and sound effects

- All music from FreeSound.org and is (CC BY 3.0)
- Added sound effects for looting and stabbing
- Added background music
- Added short suspense music

Fixes #70

Commit:
refactor: lots of cleanup and add splash screen

- Add flashing text scene in Scenes/FlashingText/
- Add splash scene in Scenes/Splash/
- Moved global classes to Classes/
- Implement global scene loading event
- Add music player

Fixes #53, #71

eco9898 reviewed (changes requested):
Some questions on discord. A lot has been changed

Commit:
refactor: change the way characters are created

- Create Classes/Attacks.gd to store attacks
- Create Classes/CharacterDefaults.gd to store default character stats
- Move Classes/Characters/Character.gd functions to
Classes/CharacterManager.gd and make them static
- Created Classes/DictonaryFunc.gd for useful dictonary functions
- Created Classes/Items.gd to store item stats
- Updated Battle.gd, CharacterCreation and CharacterSelection to work with changes
- SaveGame.gd: Removed character_class_to_data(Character) and
data_to_character_class(player) since this is no longer needed

Commit:
fix: get items working with new system

- Update print to Core.emit_signal
- Change (ex) Attacks.melee_attacks to Attacks.melee
- Fix func update_attacks(CharacterIndex):
- Fix print error on ok save_character(data) make_dir
- Remove if !data.has("player"): from load_character since
file structure has changed

Fixes #73

eco9898 commented on Classes/CharacterDefaults.gd:
```diff
@@ -0,0 +1,316 @@
+	
+	
+	
+	knight = [
+		"sword"
+	],
+	
+	
+	
+	
+	battle_mage = [
+		"staff"
+	],
+	
+	
+	
+	
+	berserker = [
+		"axe"
+	],
+	
+	
+	
+	
+	quick_shooter = [
+		"bow", "arrow", "dagger"
+	]
```
Where are these items actually created,
Also where are skills/attacks saved

eco9898 commented on Classes/CharacterDefaults.gd:
```diff
@@ -0,0 +1,316 @@
+		
+		kills = 0,
+		items = {
+			amour = {
```
spelt wrong

eco9898 replied:
spelt wrong (armour)

eco9898 commented on Classes/CharacterDefaults.gd:
```diff
@@ -0,0 +1,316 @@
+		
+		
+		kills = 0,
+		items = {
```
equiped items (equipment) is meant to be an inventory class

josephtheengineer replied:
See latest commit!

eco9898 commented on Classes/CharacterDefaults.gd:
```diff
@@ -0,0 +1,316 @@
+			
+			
+			
+			inventory = [
```
inventory is not a sub var of equiped items,
also it should be an inventory class

josephtheengineer replied:
See latest commit!

eco9898 commented on Classes/CharacterDefaults.gd:
```diff
@@ -0,0 +1,316 @@
+	
+	
+	
+	knight = [
+		"sword"
+	],
+	
+	
+	
+	
+	battle_mage = [
+		"staff"
+	],
+	
+	
+	
+	
+	berserker = [
+		"axe"
+	],
+	
+	
+	
+	
+	quick_shooter = [
+		"bow", "arrow", "dagger"
+	]
```
Where are the starting skills/attacks saved

eco9898 replied:
Add a classattack array to add new attacks based of the chosen class, also the amount of arrows should be specified

eco9898 commented on Classes/CharacterDefaults.gd:
```diff
@@ -0,0 +1,316 @@
+		
+		
+		kills = 0,
+		equipment = [
+			
+		],
+		inventory = [
+			
+		],
```
better, but these should be an inventory class

eco9898 commented on Classes/CharacterDefaults.gd:
```diff
@@ -0,0 +1,316 @@
+		
+		
+		
+		equipBuffs = {
+			strength = 0,
+			perception = 0,
+			endurance = 0,
+			charisma = 0,
+			intelligence = 0,
+			agility = 0,
+			luck = 0,
+			melee = 0,
+			mana = 0,
+			defense = 0
```
equipbuffs is dynamically adjusted on creation based off levels and equipment

eco9898 commented on Classes/CharacterManager.gd:
```diff
@@ -0,0 +1,115 @@
+#				if items[item.broadType][self.check(item, 1)[1]][1] == 0:
+#					items[item.broadType].remove(self.check(item, 1)[1])
+#				return [true, "Done"]
+#			else:
+#				return [false, "Doesn't Exist"]
+#	else:
+#		return [false, "Quantity Error"]
+
+## [if that many exist, index of that many items]
+static func check(item, quantity:int):
+	pass
+#	var done = false
+#	if item.broadType != "other":
+#		if !items.has(item.broadType) or !items[item.broadType].has(item.type):
+#			Core.emit_signal("msg", "Invalid item type" + str(item.broadType) + ", " + str(item.type), Core.WARN, "character_manager")
+#			return false
+#
+#		for inventory_item in items[item.broadType][item.type]:
+#			if item.isTheSameAs(inventory_item[0]) && inventory_item[1] >= quantity:
+#				return [true, items[item.broadType][item.type].find(inventory_item)]
+#	else:
+#		for inventory_item in items[item.broadType]:
+#			if item.isTheSameAs(inventory_item[0]) && inventory_item[1] >= quantity:
+#				return [true, items[item.broadType].find(inventory_item)]
+#	if !done:
+#		return [false, -1]
+#func text():
+#	return "items: " + str(items) + " money: " + str(money)
+
+static func unequip(item):
+	pass
+#	if equipment.check(item, 1)[0]:
+#		equipment.remove(item, 1)
+#		for buffKey in item.buffs.keys():
+#			equipBuffs[buffKey] -=item.buffs[buffKey]
+#		return [true, "Done"]
+#	else:
+#			return [false, "Doesn't Exist"]
+#func text():
+#	return character_name
```
This is used to manage inventory instead of using the inventory class?

eco9898 commented on Classes/CharacterManager.gd:
```diff
@@ -0,0 +1,115 @@
+#func text():
+#	return character_name
+
+static func set_level():
+	pass
+#		var tempLevel = level
+#	var count = 1
+#	while tempLevel >= 5:
+#		stats["Strength"] += 1
+#		stats["Intelligence"] += 1
+#		stats["Agility"] += 1
+#		stats["Luck"] += 1
+#		if count%3==0:
+#			stats["Perception"] += 1
+#			stats["Endurance"] += 1
+#			stats["Charisma"] += 1
+#		count+=1
+#		tempLevel-=5
```
Now that I think about it, this shouldn't directly modify stats as it will need to be processed each level.
Instead it should have its on level buffs, like equipbuffs that should be reset when called and then modified so it doesn't accidently layer up the level buffs

josephtheengineer replied:
Gotcha!

eco9898 commented on Classes/DictonaryFunc.gd:
```diff
@@ -0,0 +1,46 @@
+	for k in mapList: 
+		dataDict = dataDict[k]
+	return dataDict
+
+
+# Set a given data in a dictionary with position provided as a list
+# eg. path="dict.subdict.value"
+#     setInDict(dict, path.split(".", false), value)
+static func setInDict(dataDict, mapList, value):
+	var dict = mapList
+	dict.remove(dict.size() - 1)
+	
+	for k in dict: 
+		dataDict = dataDict[k]
+	dataDict[mapList[-1]] = value
+
+
+static func merge_dict(target, patch):
+	for key in patch:
+		if target.has(key):
+			var tv = target[key]
+			if typeof(tv) == TYPE_DICTIONARY:
+				merge_dict(tv, patch[key])
+			else:
+				target[key] = patch[key]
+		else:
+			target[key] = patch[key]
+ 
+
+static func find_all_value(_dict, target):
+	for key in _dict.keys():
+		#case 1: it is a dictionary but not match the key
+		if typeof(_dict[key]) == TYPE_DICTIONARY and key != target:
+			return find_all_value(_dict[key], target)
+		#case 2: it is a dictionary but match the key -> put it in result
+		elif typeof(_dict[key]) == TYPE_DICTIONARY and key == target:
+			return [_dict[key]] + find_all_value(_dict[key], target)
+		#case 3: it is not dictionary and match the key -> put it in result
+		elif key == target:
+			return [_dict[key]]
```
Don't know what this is for, I guess you needed it for something

eco9898 commented on Classes/Items.gd:
```diff
@@ -0,0 +1,141 @@
+	arrow = {
+		nickname = "Arrow",
+		name = "arrow",
+		type = "consumables",
+		hand = "one-handed",
+		tags = [],
+		buffs = {
+			strength = 0,
+			perception = 0,
+			endurance = 0,
+			charisma = 0,
+			intelligence = 0,
+			agility = 0,
+			luck = 0,
+		},
+		level = 1,
+		level_requirement = 1,
+	},
+	
+	
+	
+	
+	dagger = {
+		nickname = "Dagger",
+		name = "dagger",
+		type = "melee",
+		hand = "one-handed",
+		tags = [],
+		buffs = {
+			strength = 0,
+			perception = 0,
+			endurance = 0,
+			charisma = 0,
+			intelligence = 0,
+			agility = 0,
+			luck = 0,
+		},
+		level = 1,
+		level_requirement = 1,
+	}
```
While this is ok, because these are called each time a character is created, not all weapons will be hard coded. Some weapons and items will be dynamically created and leveled.

eco9898 commented on Classes/CharacterDefaults.gd:
```diff
@@ -0,0 +1,316 @@
+	#                                 
+	#                                 
+	
+	alrune = {
+		meta = {
+			name = "Alrune",
+			nickname = "",
+			info = "",
+		},
```
This character was meant to test the dynamically created characters, however now it is a good way of testing if companion characters can be loaded in

eco9898 commented on Scenes/BattleScenes/Create_Battle.gd:
```diff
@@ -1,12 +0,0 @@
-extends Node
-
-static func switchScene(object, sceneName:String, background:String, friendlies:Array, enemies:Array):
-	##Switch Scene
-	object.add_child(load("res://Scenes/BattleScenes/Battle.tscn").instance())
-	object.get_node("Battle").enemies = enemies
-	object.get_node("Battle").friendlies = friendlies
-	##Create Setting
-	object.get_node("Battle").update_Setting(sceneName, background)
-	##Update stats
-	object.get_node("Battle").update_Characters()
-	object.get_node("Battle").update_Attacks(0)
```
How is this handled now?

josephtheengineer replied:
refactored, see load_battle()

eco9898 commented on Scenes/CharacterCreation/CharacterCreation.gd:
```diff
@@ -18,33 +16,45 @@ func _ready():
+	spec[2].get_node("Labels/Label").text = "Endurace"
+	spec[3].get_node("Labels/Label").text = "Charisma"
+	
+	var ial = get_node("MainMenu/Choices/Stats/Display/Menu/IAL").get_children()
+	ial[0].get_node("Labels/Label").text = "Intelligence"
+	ial[1].get_node("Labels/Label").text = "Agility"
+	ial[2].get_node("Labels/Label").text = "Luck"
 #	get_node("MainMenu/Choices/ProfileSelection/Profiles").get_children()[selectedCharacter].get_node("Border").show()
 #	get_node("MainMenu/Choices/ProfileSelection/Profiles").get_children()[selectedCharacter].chosen = true
-	get_node("MainMenu/Choices/Equipment/Classes/Knight").get_node("ClassName").text = "Knight"
-	get_node("MainMenu/Choices/Equipment/Classes/Knight").get_node("ImgCenter/ClassImg").texture = load("res://Assets/Images/Icons/Classes/Badge_warrior.png")
-	get_node("MainMenu/Choices/Equipment/Classes/Berserker").get_node("ClassName").text = "Berserker"
-	get_node("MainMenu/Choices/Equipment/Classes/Berserker").get_node("ImgCenter/ClassImg").texture = load("res://Assets/Images/Icons/Classes/Badge_barbarian.png")
-	get_node("MainMenu/Choices/Equipment/Classes/BattleMage").get_node("ClassName").text = "Battle Mage"
-	get_node("MainMenu/Choices/Equipment/Classes/BattleMage").get_node("ImgCenter/ClassImg").texture = load("res://Assets/Images/Icons/Classes/Badge_mage.png")
-	get_node("MainMenu/Choices/Equipment/Classes/QuickShooter").get_node("ClassName").text = "Quick Shooter"
-	get_node("MainMenu/Choices/Equipment/Classes/QuickShooter").get_node("ImgCenter/ClassImg").texture = load("res://Assets/Images/Icons/Classes/Badge_hunter.PNG")
+	var knight = get_node("MainMenu/Choices/Equipment/Classes/Knight")
+	knight.get_node("ClassName").text = "Knight"
+	knight.get_node("ImgCenter/ClassImg").texture = load("res://Assets/Images/Icons/Classes/Badge_warrior.png")
+	
+	var berserker = get_node("MainMenu/Choices/Equipment/Classes/Berserker")
+	berserker.get_node("ClassName").text = "Berserker"
+	berserker.get_node("ImgCenter/ClassImg").texture = load("res://Assets/Images/Icons/Classes/Badge_barbarian.png")
+	
+	var battle_mage = get_node("MainMenu/Choices/Equipment/Classes/BattleMage")
+	battle_mage.get_node("ClassName").text = "Battle Mage"
+	battle_mage.get_node("ImgCenter/ClassImg").texture = load("res://Assets/Images/Icons/Classes/Badge_mage.png")
+	
+	var quick_shooter = get_node("MainMenu/Choices/Equipment/Classes/QuickShooter")
+	quick_shooter.get_node("ClassName").text = "Quick Shooter"
+	quick_shooter.get_node("ImgCenter/ClassImg").texture = load("res://Assets/Images/Icons/Classes/Badge_hunter.PNG")
```
What was changed here?
Just making the dynamic fields pre named?

josephtheengineer replied:
Yep! Refactoring work

eco9898 commented on Scenes/CharacterCreation/CharacterCreation.gd:
```diff
@@ -71,59 +81,44 @@ func _on_Name_text_changed():
+	
+	var player2 = CharacterManager.create(character_names[selected_character])
+
+	var enemy = CharacterManager.create("death_hound")
+	var enemy2 = CharacterManager.create("death_hound")
+	var enemy3 = CharacterManager.create("death_hound")
```
A level will need to be passed on creation to scale it

eco9898 replied:
Also, after being created any extra skills that may need to be added will be added seperately?

eco9898 replied:
So this can be done by the character management script?

josephtheengineer replied:
Yep!

josephtheengineer replied:
I'll save this for another PR!

eco9898 commented on Scenes/CharacterSelection/CharacterSelection.gd:
```diff
@@ -21,37 +23,61 @@ onready var SaveGame = get_node("/root/Variables").SaveGame
-		ButtonInstance.get_node("VBox/Content/HBox/CenterProfile/HBox/Profile").flip_v = player.pic[1][1]
+		ButtonInstance.get_node("VBox/Content/HBox/CenterProfile/HBox/Profile").texture = load(
+			"res://Assets/Images/Profiles/" + picture.path)
+		#ButtonInstance.get_node("VBox/Content/HBox/CenterProfile/HBox/Profile").flip_h = player.pic[1][0]
+		#ButtonInstance.get_node("VBox/Content/HBox/CenterProfile/HBox/Profile").flip_v = player.pic[1][1]
```
The image still needs to be flipped right?

josephtheengineer replied:
Hmmmm, I think so

eco9898 replied:
So why is it commented out?

josephtheengineer replied:
I don't think profile images display correctly at all in the selection menu

josephtheengineer replied:
Ok it works now! :D

eco9898 commented on Scenes/BattleScenes/Battle.gd:
```diff
@@ -1,15 +1,57 @@
 extends Node
+class_name Battle
+const script_name := "battle"
+
```
Is the changes to this file just compatiability and renaming?

josephtheengineer replied:
script_name is used for logging and class_name is used for instancing although in this case it's unneeded (since it's only used in it's specific scene). I'll remove this in the next commit!

Commit:
revert: fix inventory items

- Add Item.gd
- Add Inventory.gd

Commit:
fix: profile images do not load

- Add "res://Assets/Images/Profiles/" to picture path

Fixes #74

eco9898 commented on Classes/CharacterManager.gd:
```diff
@@ -0,0 +1,115 @@
+#				if items[item.broadType][self.check(item, 1)[1]][1] == 0:
+#					items[item.broadType].remove(self.check(item, 1)[1])
+#				return [true, "Done"]
+#			else:
+#				return [false, "Doesn't Exist"]
+#	else:
+#		return [false, "Quantity Error"]
+
+## [if that many exist, index of that many items]
+static func check(item, quantity:int):
+	pass
+#	var done = false
+#	if item.broadType != "other":
+#		if !items.has(item.broadType) or !items[item.broadType].has(item.type):
+#			Core.emit_signal("msg", "Invalid item type" + str(item.broadType) + ", " + str(item.type), Core.WARN, "character_manager")
+#			return false
+#
+#		for inventory_item in items[item.broadType][item.type]:
+#			if item.isTheSameAs(inventory_item[0]) && inventory_item[1] >= quantity:
+#				return [true, items[item.broadType][item.type].find(inventory_item)]
+#	else:
+#		for inventory_item in items[item.broadType]:
+#			if item.isTheSameAs(inventory_item[0]) && inventory_item[1] >= quantity:
+#				return [true, items[item.broadType].find(inventory_item)]
+#	if !done:
+#		return [false, -1]
+#func text():
+#	return "items: " + str(items) + " money: " + str(money)
+
+static func unequip(item):
+	pass
+#	if equipment.check(item, 1)[0]:
+#		equipment.remove(item, 1)
+#		for buffKey in item.buffs.keys():
+#			equipBuffs[buffKey] -=item.buffs[buffKey]
+#		return [true, "Done"]
+#	else:
+#			return [false, "Doesn't Exist"]
+#func text():
+#	return character_name
```
Should be added back into the character class or add a character argument

eco9898 commented on Classes/CharacterManager.gd:
```diff
@@ -0,0 +1,115 @@
+
+static func create(character):
+	return CharacterDefaults.new().characters[character]
+
```
Add a function to learn attacks instead of directly adding attacks to an array

eco9898 replied:
Also add a function that calls inventory.add and inventory.remove instead of directly accessing it to make things handled the same

eco9898 commented on Classes/Item.gd:
```diff
@@ -0,0 +1,25 @@
+var type = ""
+var name = ""
+var typeStat = 0
+var buffs = {"Strength":0, "Perception":0, "Endurance":0, "Charisma":0, "Intelligence":0, "Agility":0, "Luck":0}
```
damage and defense variables missing

eco9898 commented on Classes/Items.gd:
```diff
@@ -0,0 +1,141 @@
+	
+	
+	
+	sword = {
+		nickname = "Sword",
+		name = "sword",
+		type = "melee",
+		hand = "one-handed",
+		tags = [],
+		buffs = {
+			strength = 0,
+			perception = 0,
+			endurance = 0,
+			charisma = 0,
+			intelligence = 0,
+			agility = 0,
+			luck = 0,
+		},
+		level = 1,
+		level_requirement = 1,
+	},
```
Invalid data for creating the items

Commit:
fix: item equip now works again

- Reverted the item adding changes to CharacterCreation.gd
- Changed Item _init buffs to a Dictionary

eco9898 commented on Classes/CharacterDefaults.gd:
```diff
@@ -0,0 +1,292 @@
+		
+		
+		kills = 0,
+		items = {
```
This needs to be removed

eco9898 commented on Scenes/CharacterCreation/CharacterCreation.gd:
```diff
@@ -95,7 +95,55 @@ func _on_Complete_pressed():
+			CharacterManager.equip(player, sword)
+			print (player.equip_buffs)
+			player.attacks.melee.append("strike")
+		"battle_mage":
+			var buffs = Dictionary()
+			buffs.intelligence = 2
+			buffs.mana_attack = 10
+			var staff = Item.new("weapons", "magic", "staff", "Staff", buffs, 1)
+			player.items.inventory.add(staff, 1)
+			CharacterManager.equip(player, staff)
+			player.attacks.mana.append("flame")
+		"berserker":
+			var buffs = Dictionary()
+			buffs.strength = 3
+			buffs.agility = -1
+			buffs.melee_attack = 20
+			var axe = Item.new("weapons", "melee", "two-handed axe", "Two-handed Battle Axe", buffs, 1)
+			player.items.inventory.add(axe, 1)
+			CharacterManager.equip(player, axe)
+			player.attacks.melee.append("strike")
+		"quick_shooter":
+			var buffs = Dictionary()
+			buffs.agility = 2
+			var bow = Item.new("weapons", "ranged", "hunting bow", "Bow", buffs, 1)
+			var buffs2 = Dictionary()
+			buffs2.luck = 1
+			buffs2.melee_attack = 5
+			var arrow = Item.new("weapons", "consumables", "arrow", "Arrow", buffs2, 1)
+			var buffs3 = Dictionary()
+			buffs3.strength = 1
+			buffs3.melee_attack = 5
+			var dagger = Item.new("weapons", "melee", "dagger", "Dagger", buffs3, 1)
+			player.items.inventory.add(bow, 1)
+			CharacterManager.equip(player, bow)
+			player.items.inventory.add(arrow, 10)
+			CharacterManager.equip(player, arrow)
+			player.items.inventory.add(dagger, 1)
+			CharacterManager.equip(player, arrow)
+			player.items.attacks.ranged.append("quick_shot")
+			player.items.attacks.melee.append("strike")
```
Change this to reference a defaults dictionary for what should be added to the inventory and attack list

eco9898 replied:
address inventory and attack changes by characterManager

eco9898 replied:
Character manager should have a load class function which will add and equip items as well as adding attacks

Commit:
refactor: hard code in class attributes

- Add const starting_attacks to CharacterDefaults.gd
- Item buffs are now calcuated when needed and not stored (no reduct
complexity)
- Items and Inventories are now stored as pure data
- Functions moved to InventoryManager and ItemManager
- Refactored InventoryManager
- Updated item buffs to correct values
- Implement function load_class(character, character_class)
- Changed 'weapon' type for items to a more general 'equipment' type
- Replaced CharacterCreation's match chosen_equip with
CharacterManager.load_class(player, chosen_equip)

Commit:
feature: add a function to learn attacks

- Replace character.attacks[attack_type].append(attack) to
  learn_attack(character_data, attack_type, attack_name)
- Also rename character -> character_data in load_class
- Also remove unneeded argument 'item' from calcuate_item_buffs()

Commit:
fix: equiped items are now stored by their index

- Change var characters to const characters in CharacterDefaults
- Refactor CharacterManager
- Make InventoryManager work on a specific inventory and have
CharacterManager wrap that

Fixes #78
Works on #77 #76

Commit:
refactor: remove unused global class Battle

Commit:
feature: add wooden_ui assets, fog and sound fx

- Add Assets/Images/Wooden_UI
- Add various Assets/Sounds/Effects
- Add various Assets/Sounds/Music

AudioManager
- Create AudioManager for easily creating audio
  players

CharacterManager.gd
- remove unneeded whitespace

AudioDebugScene
- Create AudioDebugScene for testing the audio buses

CharacterCreation.gd
- add support for flashing text button

CharacterCreation.tscn
- Add fog background
- Make button text flash when enabled
- Change GAME_TITLE -> RPG GODOT

CharacterSelection.gd
- Add support for a profile border
- Fix crash when selecting a character
  - change saveFile -> file

CharacterSelection.tscn
- Make selection buttons different colors (Wooden_UI/frame)
- Add banner to the top of the scene where the title is

CharacterSelectionButton.tscn
- Add profile border
- Change button background to Wooden_UI/bg_01_02_landscape.png

FlashingText
- Create text script and scene that flashes when
  state is set to States.FLASHING

Loading
- Make loading show the script_name and level like in Core
- Add 1 sec wait
- Make loading play Music via the AudioManager

Overlay
- Create an Overlay / Background element that contains
  some cool animations
- Overlay runs lightning storm on splash screen
- The lightning storm uses a 2d light and some sound effects
- Overlay animates the fog
- Overlay renders some 2d particles Over top the screen for
  a cool depth effect

Splash
- Add script name to splash
- Make splash call start-storm on Overlay
- Splash tells the flashing text to show after the logo
  text has faded in

Shaders/noise.shader
- Add generic noise shader (used for fog)

AudioBus
- Create default_bus_layout.tres to seperate and control the
  volume of different sounds
- Some buses also have different sound effects / filters

layer_names
- Add layers for 2d physics

eco9898 commented on Scenes/BattleScenes/Battle.gd:
```diff
@@ -1,15 +1,56 @@
 extends Node
```
What was actually changed here, you moved a lot around

josephtheengineer replied:
I removed class_name Battle from the battle script because we didn't use it

eco9898 commented on Classes/CharacterManager.gd:
```diff
@@ -0,0 +1,73 @@
+static func set_level(character_data: Dictionary, level: int):
+	Core.emit_signal("msg", "This function is not implemented yet!", Core.WARN, "character_manager")
+#		var tempLevel = level
+#	var count = 1
+#	while tempLevel >= 5:
+#		stats["Strength"] += 1
+#		stats["Intelligence"] += 1
+#		stats["Agility"] += 1
+#		stats["Luck"] += 1
+#		if count%3==0:
+#			stats["Perception"] += 1
+#			stats["Endurance"] += 1
+#			stats["Charisma"] += 1
+#		count+=1
+#		tempLevel-=5
```
This is only for the hounds, not all of them

eco9898 replied:
It is special for each enemy/friendly type

josephtheengineer replied:
I don't think so!

Commit:
refactor: move Core log to Log, rename BattleScene

Log
- Move Core functions responsible for logging to a
  seperate Log class
- Change all references to Core.LEVEL to Log.LEVEL

Battle
- Rename BattleScenes folder to Battle because the
  parent is also call scenes

Starting Equipment
- Make items a dictonary and allow a quantity field
- This allows us to easily give the the player 15
  arrows with 'quantity = 15'
- Reflect this change in CharacterManager

CharacterManager
- Implement set_level and process_level

InventoryManager
- Fix an issue where items would no get their quantity
  set if the item doesn't already exist

  add 'item_data.quantity = quantity'
- Fix an issue where items would check different just
  becuase they have a different quantity

  Solution: remove the quantity value before comparing
            item_check_data.erase("quantity")

ItemManager
- Move function to InventoryManager and delete

eco9898 commented on Classes/CharacterManager.gd:
```diff
@@ -0,0 +1,79 @@
+	InventoryManager.remove(character_data.items.inventory, item_data, quantity)
+
+
+static func equip(character_data: Dictionary, item_index: int):
+	if !character_data.items.inventory[item_index]:
+		return false
+	var item = character_data.items.inventory[item_index]
+
+	if item.level_requirement <= character_data.level:
+		character_data.items[item.type][item.subtype].append(item_index)
+	else:
+		return false
+
+
+static func unequip(character_data: Dictionary, item_index: int):
+	if !character_data.items.inventory[item_index]:
+		return false
+	var item = character_data.items.inventory[item_index]
+	
+	character_data.items[item.type][item.subtype].erase(item_index)
```
Inventory was not meant to be a sub value of items
Inventory and items are both inventory classes, which don't have sub inventories if they did this would be an infinite inventory of inventories

josephtheengineer replied:
hmmm looks good to me?
```
{
  "meta": {
    "name": "Alrune",
    "nickname": "",
    "info": ""
  },
  "file": "",
  "stats": [
    3,
    3,
    3,
    3,
    3,
    3,
    3
  ],
  "picture": {
    "path": "Friendlies/Tex_AnimeAva_28.png",
    "border": {
      "path": "ImageBorder.png",
      "shown": true
    }
  },
  "attacks": {
    "melee": [
      "punch",
      "strike"
    ],
    "ranged": [],
    "mana": []
  },
  "level": 1,
  "skills": {},
  "ap": {
    "max": 1,
    "current": 0.5
  },
  "health": {
    "max": 100,
    "current": 100
  },
  "mana": {
    "max": 50,
    "current": 100
  },
  "kills": 0,
  "items": {
    "armour": {
      "head": [],
      "torso": [],
      "arms": [],
      "legs": [],
      "feet": [],
      "shield": []
    },
    "equipment": {
      "melee": [
        0
      ],
      "ranged": [],
      "consumables": [],
      "magic": []
    },
    "inventory": [
      {
        "nickname": "Sword",
        "name": "sword",
        "type": "equipment",
        "subtype": "melee",
        "hand": "one-handed",
        "tags": [],
        "buffs": {
          "strength": 2,
          "perception": 0,
          "endurance": 0,
          "charisma": 0,
          "intelligence": 0,
          "agility": 0,
          "luck": 0,
          "melee": 10,
          "mana": 0,
          "defense": 0
        },
        "level": 1,
        "level_requirement": 1,
        "quantity": 1
      }
    ]
  },
  "nickname": "fpg"
}
```

eco9898 replied:
items shouldn't exist
equipment and inventory are the only ones
you've mixed up items with what equipment is for
equipment is the currently equipped stuff and inventory is everything the character owns
items is not needed
and if you do want to keep it the armour will be in equipment and items will just have the two inventory classes

eco9898 replied:
also, why is 0 equiped as a weapon?
0 is not a weapon
equipment should hold an exact copy of the item otherwise the equip/unequip and battle code needs to be changed

eco9898 replied:
I thought you had set this back, but now it seems like you've gone back to what you changed it to

eco9898 commented on Classes/CharacterManager.gd:
```diff
@@ -0,0 +1,79 @@
+	Core.emit_signal("msg", "Setting character level to " + str(character_data.level), Log.INFO, "character_manager")
+	character_data.level = level
+
+static func process_level(character_data: Dictionary):
+	Core.emit_signal("msg", "Processing character level " + str(character_data.level) + " stats", Log.INFO, "character_manager")
+	var temp_level = character_data.level
+	var count = 1
+	var new_character_data = character_data
+	while temp_level >= 5:
+		new_character_data.stats.strength += 1
+		new_character_data.stats.intelligence += 1
+		new_character_data.stats.agility +=1
+		new_character_data.stats.luck += 1
+		
+		if count%3==0:
+			new_character_data.stats.perception += 1
+			new_character_data.stats.endurance += 1
+			new_character_data.stats.charisma += 1
+	count+=1
+	temp_level-=5
+	return new_character_data.stats
```
um, this won't work properly

step 1. level buffs had to be it's own stats like equip buffs
step 2. level buffs are cleared before calculating
step 3. the new value is created

Also side note, a class value needs to be given because that leveling is only for the hound.

eco9898 replied:
also, count and temp level were meant to be in the while loop
look at the original code
It won't work if you modify those values after they are meant to be checked
how is this suppose to leave the while loop?

josephtheengineer replied:
woops!

josephtheengineer replied:
Ok I think I fixed it!
```
{
  "meta": {
    "name": "Alrune",
    "nickname": "",
    "info": ""
  },
  "file": "Alrune - 1592107128",
  "stats": {
    "strength": 3,
    "perception": 3,
    "endurance": 3,
    "charisma": 3,
    "intelligence": 3,
    "agility": 3,
    "luck": 3,
    "melee": 0,
    "mana": 0,
    "defense": 0
  },
  "picture": {
    "path": "Friendlies/Tex_AnimeAva_28.png",
    "border": {
      "path": "ImageBorder.png",
      "shown": true
    }
  },
  "attacks": {
    "melee": [
      "punch"
    ],
    "ranged": [
      "quick_shot",
      "strike"
    ],
    "mana": []
  },
  "level": 10,
  "skills": {},
  "ap": {
    "max": 1,
    "current": 0.5
  },
  "health": {
    "max": 100,
    "current": 100
  },
  "mana": {
    "max": 50,
    "current": 100
  },
  "kills": 0,
  "items": {
    "armour": {
      "head": [],
      "torso": [],
      "arms": [],
      "legs": [],
      "feet": [],
      "shield": []
    },
    "equipment": {
      "melee": [
        2
      ],
      "ranged": [
        0
      ],
      "consumables": [
        1
      ],
      "magic": []
    },
    "inventory": [
      {
        "nickname": "Hunting bow",
        "name": "bow",
        "type": "equipment",
        "subtype": "ranged",
        "hand": "two-handed",
        "tags": [],
        "buffs": {
          "strength": 0,
          "perception": 0,
          "endurance": 0,
          "charisma": 0,
          "intelligence": 0,
          "agility": 2,
          "luck": 0,
          "melee": 0,
          "mana": 0,
          "defense": 0
        },
        "level": 1,
        "level_requirement": 1,
        "quantity": 1
      },
      {
        "nickname": "Arrow",
        "name": "arrow",
        "type": "equipment",
        "subtype": "consumables",
        "hand": "one-handed",
        "tags": [],
        "buffs": {
          "strength": 0,
          "perception": 0,
          "endurance": 0,
          "charisma": 0,
          "intelligence": 0,
          "agility": 0,
          "luck": 1,
          "melee": 4,
          "mana": 0,
          "defense": 0
        },
        "level": 1,
        "level_requirement": 1,
        "quantity": 15
      },
      {
        "nickname": "Dagger",
        "name": "dagger",
        "type": "equipment",
        "subtype": "melee",
        "hand": "one-handed",
        "tags": [],
        "buffs": {
          "strength": 0,
          "perception": 0,
          "endurance": 0,
          "charisma": 0,
          "intelligence": 0,
          "agility": 0,
          "luck": 1,
          "melee": 5,
          "mana": 0,
          "defense": 0
        },
        "level": 1,
        "level_requirement": 1,
        "quantity": 1
      }
    ]
  },
  "nickname": "fpg",
  "level_stats": {
    "strength": 2,
    "perception": 0,
    "endurance": 0,
    "charisma": 0,
    "intelligence": 2,
    "agility": 2,
    "luck": 2,
    "melee": 0,
    "mana": 0,
    "defense": 0
  }
}
```

eco9898 replied:
Do we need to save this: "file": "Alrune - 1592107128"?
Wouldn't it just be saved by the environment not the player

eco9898 replied:
wasn't equipment and items just meant to be inventory.new()?

eco9898 replied:
Or is this more like a save data to load a character from?

josephtheengineer replied:
See commit d15061d38c7f4b452611f5fe3e5457b7dc30d1ae

(HEAD -> develop, origin/develop, github/develop)
Author: josephtheengineer <joseph@theengineer.life>
Date:   Fri Jun 12 17:36:31 2020 +1000

    refactor: hard code in class attributes
    
    - Add const starting_attacks to CharacterDefaults.gd
    - Item buffs are now calcuated when needed and not stored (no reduct
    complexity)
    - Items and Inventories are now stored as pure data
    - Functions moved to InventoryManager and ItemManager
    - Refactored InventoryManager
    - Updated item buffs to correct values
    - Implement function load_class(character, character_class)
    - Changed 'weapon' type for items to a more general 'equipment' type
    - Replaced CharacterCreation's match chosen_equip with
    CharacterManager.load_class(player, chosen_equip)

josephtheengineer replied:
atm, the 'file' attribute is used to save the character :D

Commit:
fix: logging now works again, renamed Characters

Logging
- Make the logger a preload var
- Add logger.init() to _ready in Core
- Add script_name in Log
- Move Log message to after the signal gets setup

CharacterDefaults.gd -> Characters.gd
- Rename CharacterDefaults to Characters to keep
in line with other names, eg, Attacks, Items

CharacterCreation
- Fixed a bug where after creating a character
  your stats would be set to an array rather then
  a Dictonary

process_level_buffs
- Now takes a character_class argument
- process_level_buffs() is now ran when
  set_level() is run

Fixed infinite loop bug
- Put process_level_buffs count and temp_level changes
  inside the loop, woops

eco9898 commented on Scenes/BattleScenes/Battle.gd:
```diff
@@ -212,4 +229,6 @@ func update_Characters():
 			enemyPanel.get_node("VBox/Control/Blood").hide()
 		else:
+			# Enemy die
+			Core.emit_signal('msg', 'An enemy has died!', Core.INFO, self)
```
Not all enemies or friends will die

eco9898 commented on Scenes/BattleScenes/Battle.gd:
```diff
@@ -157,6 +171,7 @@ func _on_Attack_pressed():
 	AttackList.show()
 	enemies[0].health-=10
 	update_Attacks(activeCharacterIndex)
+	Core.emit_signal('msg', 'You dealt ' + str(10) + ' damage!', Core.INFO, self)
```
this will need to be changed as it was only a debug feature
